src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -425,162 +426,95 @@ internal void CheckConstraintLanguageVersionAndRuntimeSupportForConversion(Synta
 
-            MethodSymbol? getSpanConstructor(CSharpSyntaxNode syntax, TypeSymbol spanType, WellKnownMember spanMember, BindingDiagnosticBag diagnostics)
-            {
-                return (MethodSymbol?)GetWellKnownTypeMember(spanMember, diagnostics, syntax: syntax)?.SymbolAsMember((NamedTypeSymbol)spanType);
+                var result = BindCollectionInitializerCollectionLiteral(
+                    node,
+                    GetWellKnownType(WellKnownType.System_Collections_Generic_List_T, diagnostics, syntax).Construct(elementType),
+                    wasCompilerGenerated: wasCompilerGenerated,
+                    diagnostics);
+                return result.Update(result.Placeholder, result.CollectionCreation, result.Elements, targetType);
             }
 
-            bool collectionTypeImplementsIEnumerable(TypeSymbol targetType, CSharpSyntaxNode syntax, BindingDiagnosticBag diagnostics)
+            var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
+            var builder = ArrayBuilder<BoundExpression>.GetInstance(elements.Length);
+            foreach (var element in elements)
             {
-                // This implementation differs from CollectionInitializerTypeImplementsIEnumerable().
-                // That method checks for an implicit conversion from IEnumerable to the collection type,
-                // but that would allow: Nullable<StructCollection> s = [];
-                var ienumerableType = GetSpecialType(SpecialType.System_Collections_IEnumerable, diagnostics, syntax);
-                var allInterfaces = targetType is TypeParameterSymbol typeParameter
-                    ? typeParameter.AllEffectiveInterfacesNoUseSiteDiagnostics
-                    : targetType.AllInterfacesNoUseSiteDiagnostics;
-                return allInterfaces.Any(static (a, b) => a.Equals(b, TypeCompareKind.AllIgnoreOptions), ienumerableType);
+                builder.Add(convertArrayElement(element, elementType, diagnostics));
             }
+            return new BoundCollectionLiteralExpression(
+                syntax,
+                implicitReceiver,
+                collectionCreation: null,
+                builder.ToImmutableAndFree(),
+                targetType)
+            { WasCompilerGenerated = wasCompilerGenerated };
 
             BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementType, BindingDiagnosticBag diagnostics)
             {
-                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
                 var conversion = Conversions.ClassifyImplicitConversionFromExpression(element, elementType, ref useSiteInfo);
                 diagnostics.Add(element.Syntax, useSiteInfo);
                 bool hasErrors = !conversion.IsValid;
                 if (hasErrors)
```

Comment:
This is likely not a good enough check. A conversion can be valid and have use-site diagnostics.

Reply:
We're reporting the use-site errors above. Should that affect `hasErrors` though?

Reply:
Ah, I must have missed that. Thanks.